lib/Scheduler/graph_trans.cpp

```diff
@@ -2,11 +2,71 @@
 #include "opt-sched/Scheduler/bit_vector.h"
 #include "opt-sched/Scheduler/logger.h"
 #include "opt-sched/Scheduler/register.h"
+#include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
+#include <vector>
 #include <list>
 
 using namespace llvm::opt_sched;
 
+namespace {
+// Gets the registers which may be lengthened by scheduling nodeB after nodeA,
+// as compared to if nodeB was scheduled before nodeA
+llvm::SmallVector<Register const *, 10> possiblyLengthenedIfAfterOther(
+    SchedInstruction const *nodeB, llvm::ArrayRef<const Register *> bUses,
+    SchedInstruction const *nodeA, llvm::ArrayRef<const Register *> aUses) {
+  llvm::SmallVector<Register const *, 10> result;
+
+  llvm::copy_if(bUses, std::back_inserter(result), [&](const Register *useB) {
+    // Is this register also used by A?
+    // If so, reordering A and B would have no effect on this register's
+    // live range.
+    const bool usedByA =
+        llvm::any_of(aUses, [&](const Register *useA) { return useA == useB; });
+    // If this register isn't used by A, is it at least used
+    // by some successor? If so, reordering A and B would have no effect on
+    // this register's live range, as it must live until C.
+    const auto usedByC = [&] {
+      return llvm::any_of(
+          useB->GetUseList(), [&](SchedInstruction const *user) {
+            // Given: [... B ... A ...]
+            // We need to prove that the register `useB` won't be used by an
+            // instruction before A but after B. In the hypothetical schedule we
+            // are considering, A currently appears after B. Thus, it is
+            // sufficient to show that this register has a user C that is a
+            // successor of A.
+            //
+            // This is more relaxed than showing that C is a successor of B, as
+            // RcrsvScsr(B) is a subset of RcrsvScsr(A).
+            return user != nodeB &&
+                   nodeA->IsRcrsvScsr(const_cast<SchedInstruction *>(user));
+          });
+    };
+
+    return !usedByA && !usedByC();
+  });
+
+  return result;
+}
+
+// Converts a class type and a function signature into a pointer to member
+// function.
+// Usage: memfn_t<Cls, int(double) const> ==> int (Cls::*)(double) const
+template <typename Cls, typename FnSig> using memfn_t = FnSig Cls::*;
+
+// Gets the Uses or Defs for the given SchedInstruction.
+template <std::size_t N,
+          memfn_t<SchedInstruction, int16_t(Register **&)> UsesOrDefsFn>
+llvm::SmallVector<Register const *, N> getSet(SchedInstruction *node) {
+  Register **uses;
+  const int useCount = (node->*UsesOrDefsFn)(uses);
+  // Call the (Iter, Iter) constructor.
+  return {uses, uses + useCount};
```

Comment:
It seems unfortunate we need to constructing vector here for every node's uses. Can we change the underlying type to be a vector, or is that a separate patch?

Reply:
Somehow, I thought the code was previously producing a vector, but apparently not. I think I'll switch this to an `llvm::ArrayRef` to avoid the copy for now.

Yes, changing the underlying type of the use/def to be a vector is on my radar, but it's probably going to be a while before we get there.